### ISSUE TITLE
Bionic seccomp issue

### ISSUE THREAD

**Issue by rishirsinha** [cloud-hypervisor/cloud-hypervisor#2268]
I had posted a discussion a few days ago about being unable to have a VM launch using cloud hypervisor. 

I was running on a Ubuntu Bionic host with kernel version 4.15.0-135-generic.

````
./cloud-hypervisor/target/release/cloud-hypervisor --kernel ./hypervisor-fw --disk path=focal-server-cloudimg-amd64-custom.raw --cpus boot=4 --memory size=1024M --net "tap=,mac=,ip=,mask=" --rng -vv --log-file=./vmm.log --console off --serial tty
````

The output is the following:

````
Cloud Hypervisor Guest
        API server: /run/user/1000/cloud-hypervisor.13724
        vCPUs: 4
        Memory: 1024 MB
        Kernel: Some(KernelConfig { path: "./hypervisor-fw" })
        Initramfs: None
        Kernel cmdline: 
        Disk(s): Some([DiskConfig { path: Some("focal-server-cloudimg-amd64-custom.raw"), readonly: false, direct: false, iommu: false, num_queues: 1, queue_size: 128, vhost_user: false, vhost_socket: None, poll_queue: true, id: None, disable_io_uring: false }])

Booting via PVH Boot Protocol
Setting up 4 GiB identity mapping
Page tables setup
Found PCI device vendor=8086 device=d57 in slot=0
Found PCI device vendor=1af4 device=1042 in slot=1
Found PCI device vendor=1af4 device=1041 in slot=2
Found PCI device vendor=1af4 device=1044 in slot=3
PCI Device: 0:1.0 1af4:1042
Bar: type=MemorySpace32 address=e7f80000
Bar: type=MemorySpace32 address=0
Bar: type=MemorySpace32 address=0
Bar: type=MemorySpace32 address=0
Bar: type=MemorySpace32 address=0
Bar: type=Unused address=0
Virtio block device configured. Capacity: 4612096 sectors
Found EFI partition
Filesystem ready
Error loading default entry: FileError(NotFound)
Using EFI boot.
Found bootloader (BOOTX64.EFI)
Executable loaded
TPM logging failed: Unsupported
error: can't find command `hwmatch'.
[    0.000000] Linux version 5.4.0-39-generic (buildd@lcy01-amd64-016) (gcc version 9.3.0 (Ubuntu 9.3.0-10ubuntu2)) #43-Ubuntu SMP Fri Jun 19 10:28:31 UTC 2020 (Ubuntu 5.4.0-39.43-generic 5.4.41)
[    0.000000] Command line: BOOT_IMAGE=/boot/vmlinuz-5.4.0-39-generic root=PARTUUID=c1789075-d188-466c-b2a8-af2a999514f9 ro console=tty1 console=ttyS0 panic=-1
[    0.000000] KERNEL supported cpus:
[    0.000000]   Intel GenuineIntel
[    0.000000]   AMD AuthenticAMD
[    0.000000]   Hygon HygonGenuine
[    0.000000]   Centaur CentaurHauls
[    0.000000]   zhaoxin   Shanghai  
[    0.000000] x86/fpu: Supporting XSAVE feature 0x001: 'x87 floating point registers'
[    0.000000] x86/fpu: Supporting XSAVE feature 0x002: 'SSE registers'
[    0.000000] x86/fpu: Supporting XSAVE feature 0x004: 'AVX registers'
[    0.000000] x86/fpu: xstate_offset[2]:  576, xstate_sizes[2]:  256
[    0.000000] x86/fpu: Enabled xstate features 0x7, context size is 832 bytes, using 'standard' format.
[    0.000000] BIOS-provided physical RAM map:
[    0.000000] BIOS-e820: [mem 0x0000000000000000-0x000000000009ffff] usable
[    0.000000] BIOS-e820: [mem 0x0000000000100000-0x00000000001fffff] reserved
[    0.000000] BIOS-e820: [mem 0x0000000000200000-0x000000003fffffff] usable
[    0.000000] NX (Execute Disable) protection: active
[    0.000000] efi: EFI v2.70 by TPM logging failed: Unsupported
[    0.000000] 
[    0.000000] efi:  ACPI 2.0=0xa0000 
[    0.000000] secureboot: Secure boot disabled
[    0.000000] DMI not present or invalid.
[    0.000000] Hypervisor detected: KVM
[    0.000000] kvm-clock: Using msrs 4b564d01 and 4b564d00
[    0.000000] kvm-clock: cpu 0, msr e001001, primary cpu clock
[    0.000000] kvm-clock: using sched offset of 5407161465 cycles
[    0.000002] clocksource: kvm-clock: mask: 0xffffffffffffffff max_cycles: 0x1cd42e4dffb, max_idle_ns: 881590591483 ns
[    0.000003] tsc: Detected 3491.912 MHz processor
[    0.000139] last_pfn = 0x40000 max_arch_pfn = 0x400000000
[    0.000175] x86/PAT: Configuration [0-7]: WB  WC  UC- UC  WB  WP  UC- WT  
[    0.000303] found SMP MP-table at [mem 0x000f0090-0x000f009f]
[    0.000347] check: Scanning 1 areas for low memory corruption
[    0.000382] Using GB pages for direct mapping
[    0.000451] secureboot: Secure boot disabled
[    0.000463] ACPI: Early table checksum verification disabled
[    0.000467] ACPI: RSDP 0x00000000000A0000 000024 (v02 CLOUDH)
[    0.000470] ACPI: XSDT 0x00000000000A1908 00003C (v01 CLOUDH CHXSDT   00000001 CLDH 00000000)
[    0.000475] ACPI: FACP 0x00000000000A1756 000114 (v06 CLOUDH CHFACP   00000001 CLDH 00000000)
[    0.000480] ACPI: DSDT 0x00000000000A0024 001732 (v06 CLOUDH CHDSDT   00000001 CLDH 00000000)
[    0.000483] ACPI: APIC 0x00000000000A186A 000062 (v05 CLOUDH CHMADT   00000001 CLDH 00000000)
[    0.000487] ACPI: MCFG 0x00000000000A18CC 00003C (v01 CLOUDH CHMCFG   00000001 CLDH 00000000)
[    0.000799] No NUMA configuration found
[    0.000800] Faking a node at [mem 0x0000000000000000-0x000000003fffffff]
[    0.000808] NODE_DATA(0) allocated [mem 0x3ffd5000-0x3fffffff]
[    0.000977] Zone ranges:
[    0.000978]   DMA      [mem 0x0000000000001000-0x0000000000ffffff]
[    0.000979]   DMA32    [mem 0x0000000001000000-0x000000003fffffff]
[    0.000980]   Normal   empty
[    0.000980]   Device   empty
[    0.000981] Movable zone start for each node
[    0.000984] Early memory node ranges
[    0.000984]   node   0: [mem 0x0000000000001000-0x000000000009ffff]
[    0.000985]   node   0: [mem 0x0000000000200000-0x000000003fffffff]
[    0.000990] Zeroed struct page in unavailable ranges: 353 pages
[    0.000991] Initmem setup node 0 [mem 0x0000000000001000-0x000000003fffffff]
[    0.003854] ACPI: PM-Timer IO Port: 0xb008
[    0.004384] IOAPIC[0]: apic_id 0, version 17, address 0xfec00000, GSI 0-23
[    0.004388] ACPI: INT_SRC_OVR (bus 0 bus_irq 4 global_irq 4 dfl dfl)
[    0.004391] Using ACPI (MADT) for SMP configuration information
[    0.004393] smpboot: Allowing 4 CPUs, 0 hotplug CPUs
[    0.004408] PM: Registered nosave memory: [mem 0x00000000-0x00000fff]
[    0.004409] PM: Registered nosave memory: [mem 0x000a0000-0x000fffff]
[    0.004409] PM: Registered nosave memory: [mem 0x00100000-0x001fffff]
[    0.004411] [mem 0x40000000-0xffffffff] available for PCI devices
[    0.004412] Booting paravirtualized kernel on KVM
[    0.004414] clocksource: refined-jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 7645519600211568 ns
[    0.004418] setup_percpu: NR_CPUS:8192 nr_cpumask_bits:4 nr_cpu_ids:4 nr_node_ids:1
[    0.004601] percpu: Embedded 54 pages/cpu s184320 r8192 d28672 u524288
[    0.004623] KVM setup async PF for cpu 0
[    0.004627] kvm-stealtime: cpu 0, msr 3ea2c040
[    0.004631] PV qspinlock hash table entries: 256 (order: 0, 4096 bytes, linear)
[    0.004635] Built 1 zonelists, mobility grouping on.  Total pages: 257677
[    0.004636] Policy zone: DMA32
[    0.004637] Kernel command line: BOOT_IMAGE=/boot/vmlinuz-5.4.0-39-generic root=PARTUUID=c1789075-d188-466c-b2a8-af2a999514f9 ro console=tty1 console=ttyS0 panic=-1
[    0.004858] Dentry cache hash table entries: 131072 (order: 8, 1048576 bytes, linear)
[    0.004914] Inode-cache hash table entries: 65536 (order: 7, 524288 bytes, linear)
[    0.004984] mem auto-init: stack:off, heap alloc:on, heap free:off
[    0.007016] Memory: 995248K/1047164K available (14339K kernel code, 2397K rwdata, 4952K rodata, 2712K init, 4992K bss, 51916K reserved, 0K cma-reserved)
[    0.007023] random: get_random_u64 called from kmem_cache_open+0x2d/0x410 with crng_init=0
[    0.007117] SLUB: HWalign=64, Order=0-3, MinObjects=0, CPUs=4, Nodes=1
[    0.007127] Kernel/User page tables isolation: enabled
[    0.007139] ftrace: allocating 44484 entries in 174 pages
[    0.019008] rcu: Hierarchical RCU implementation.
[    0.019009] rcu:     RCU restricting CPUs from NR_CPUS=8192 to nr_cpu_ids=4.
[    0.019010]  Tasks RCU enabled.
[    0.019011] rcu: RCU calculated value of scheduler-enlistment delay is 25 jiffies.
[    0.019012] rcu: Adjusting geometry for rcu_fanout_leaf=16, nr_cpu_ids=4
[    0.020901] NR_IRQS: 524544, nr_irqs: 440, preallocated irqs: 0
[    0.020948] random: crng done (trusting CPU's manufacturer)
[    0.020965] Console: colour dummy device 80x25
[    0.021045] printk: console [tty1] enabled
[    0.086261] printk: console [ttyS0] enabled
[    0.086772] ACPI: Core revision 20190816
[    0.087268] APIC: Switch to symmetric I/O mode setup
[    0.092516] x2apic enabled
[    0.098857] Switched APIC routing to physical x2apic.
[    0.099482] clocksource: tsc-early: mask: 0xffffffffffffffff max_cycles: 0x3255787006b, max_idle_ns: 440795244845 ns
[    0.100714] Calibrating delay loop (skipped) preset value.. 6983.82 BogoMIPS (lpj=13967648)
[    0.101683] pid_max: default: 32768 minimum: 301
[    0.102272] LSM: Security Framework initializing
[    0.102827] Yama: becoming mindful.
[    0.103265] AppArmor: AppArmor initialized
[    0.103810] Mount-cache hash table entries: 2048 (order: 2, 16384 bytes, linear)
[    0.104681] Mountpoint-cache hash table entries: 2048 (order: 2, 16384 bytes, linear)
[    0.104712] *** VALIDATE tmpfs ***
[    0.104712] *** VALIDATE proc ***
[    0.104712] *** VALIDATE cgroup1 ***
[    0.104712] *** VALIDATE cgroup2 ***
[    0.104712] Last level iTLB entries: 4KB 1024, 2MB 1024, 4MB 1024
[    0.104712] Last level dTLB entries: 4KB 1024, 2MB 1024, 4MB 1024, 1GB 4
[    0.104712] Spectre V1 : Mitigation: usercopy/swapgs barriers and __user pointer sanitization
[    0.104712] Spectre V2 : Mitigation: Full generic retpoline
[    0.104712] Spectre V2 : Spectre v2 / SpectreRSB mitigation: Filling RSB on context switch
[    0.104712] Spectre V2 : Enabling Restricted Speculation for firmware calls
[    0.104712] Spectre V2 : mitigation: Enabling conditional Indirect Branch Prediction Barrier
[    0.104712] Spectre V2 : User space: Mitigation: STIBP via seccomp and prctl
[    0.104712] Speculative Store Bypass: Mitigation: Speculative Store Bypass disabled via prctl and seccomp
[    0.104712] MDS: Mitigation: Clear CPU buffers
[    0.104712] Freeing SMP alternatives memory: 40K
[    0.104712] smpboot: CPU0: Intel(R) Xeon(R) CPU E5-1650 v3 @ 3.50GHz (family: 0x6, model: 0x3f, stepping: 0x2)
[    0.104712] Performance Events: Haswell events, Intel PMU driver.
[    0.104712] ... version:                2
[    0.104712] ... bit width:              48
[    0.104712] ... generic registers:      4
[    0.104712] ... value mask:             0000ffffffffffff
[    0.104717] ... max period:             000000007fffffff
[    0.105341] ... fixed-purpose events:   3
[    0.105815] ... event mask:             000000070000000f
[    0.106500] rcu: Hierarchical SRCU implementation.
[    0.107584] smp: Bringing up secondary CPUs ...
[    0.108214] x86: Booting SMP configuration:
[    0.108712] .... node  #0, CPUs:      #1
[    0.083791] kvm-clock: cpu 1, msr e001041, secondary cpu clock
[    0.110188] KVM setup async PF for cpu 1
[    0.111401] kvm-stealtime: cpu 1, msr 3eaac040
[    0.112831] MDS CPU bug present and SMT on, data leak possible. See https://www.kernel.org/doc/html/latest/admin-guide/hw-vuln/mds.html for more details.
[    0.114423]  #2
[    0.083791] kvm-clock: cpu 2, msr e001081, secondary cpu clock
[    0.115083] KVM setup async PF for cpu 2
[    0.115083] kvm-stealtime: cpu 2, msr 3eb2c040
[    0.116815]  #3
[    0.083791] kvm-clock: cpu 3, msr e0010c1, secondary cpu clock
[    0.117959] KVM setup async PF for cpu 3
[    0.119469] kvm-stealtime: cpu 3, msr 3ebac040
[    0.120772] smp: Brought up 1 node, 4 CPUs
[    0.121451] smpboot: Max logical packages: 1
[    0.124720] smpboot: Total of 4 processors activated (27935.29 BogoMIPS)
[    0.126056] devtmpfs: initialized
[    0.126056] x86/mm: Memory block size: 128MB
[    0.126159] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 7645041785100000 ns
[    0.128727] futex hash table entries: 1024 (order: 4, 65536 bytes, linear)
[    0.129895] pinctrl core: initialized pinctrl subsystem
Bad system call
````

But I ran the same commands on another focal host and the VM came up. Looks like something is broken on bionic hosts. I am more than happy to help with whatever diagnostic data is needed to debug this.

**Comment by michael2012z** [comment]
The crash could be caused by seccomp violation. 

Appending argument `--seccomp log` should bypass the seccomp check, hopefully the VM can boot then. And the problematic system call will be logged in `dmesg` output.

If that works, could you attach `dmesg` printings? Thank you.

**Comment by rishirsinha** [comment]
```` --seccomp log```` does mitigate the issue. 

```` dmesg | grep seccom```` gives the following:

````
[    0.032850] Spectre V2 : User space: Mitigation: STIBP via seccomp and prctl
[    0.032851] Speculative Store Bypass: Mitigation: Speculative Store Bypass disabled via prctl and seccomp
````

**Comment by sboeuf** [comment]
Could you run Cloud-Hypervisor with `strace -f` (make sure you use `--seccomp true` in the command line), and identify which is the faulty syscall?

Also, are you running the latest Cloud-Hypervisor version? I'm asking because we very recently merged a seccomp fix here https://github.com/cloud-hypervisor/cloud-hypervisor/pull/2257

**Comment by michael2012z** [comment]
Alternative to run cloud-hypervisor with `strace -f` is checking the last lines of `dmesg` after starting cloud-hypervisor with `--seccomp log`.

The related trace should be like (it is not captured by grepping "seccomp"):

[2916019.947851] audit: type=1326 audit(1613373008.259:2603): auid=1001 uid=0 gid=0 ses=3 subj==unconfined pid=185611 comm="vmm" exe="/home/michael/ws/src/github.com/cloud-hypervisor/cloud-hypervisor/target/debug/cloud-hypervisor" sig=0 arch=c00000b7 **syscall=29** compat=0 ip=0xffff87886b1c code=0x7ffc0000

**29** is the problematic system call in my example.

**Comment by sboeuf** [comment]
If you want the log to show up, you need to make sure your host machine is running a kernel with `audit=1`.

**Comment by sboeuf** [comment]
FWIW, I tried to reproduce the issue on Azure with the following host `5.4.0-1039-azure #41~18.04.1-Ubuntu`, but I could not.

**Comment by rishirsinha** [comment]
> FWIW, I tried to reproduce the issue on Azure with the following host `5.4.0-1039-azure #41~18.04.1-Ubuntu`, but I could not.

I agree, it does not happen on kernel 5.* just on kernel 4.*. I will try to repro the rest with the detailed logs and then will report.

**Comment by rishirsinha** [comment]
> Could you run Cloud-Hypervisor with `strace -f` (make sure you use `--seccomp true` in the command line), and identify which is the faulty syscall?
> 
> Also, are you running the latest Cloud-Hypervisor version? I'm asking because we very recently merged a seccomp fix here #2257

@sboeuf , it looks like that fix has fixed the bad syscall issue I was seeing. My last sync was on Monday and so before when the fix went in. I will close the issue.

Fixed by #2257
